cloudflare/resource_cloudflare_custom_ssl.go

```diff
@@ -0,0 +1,284 @@
+package cloudflare
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/cloudflare/cloudflare-go"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/pkg/errors"
+)
+
+func resourceCloudflareCustomSsl() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudflareCustomSslCreate,
+		Read:   resourceCloudflareCustomSslRead,
+		Update: resourceCloudflareCustomSslUpdate,
+		Delete: resourceCloudflareCustomSslDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceCloudflareCustomSslImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"custom_ssl_priority": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"priority": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"custom_ssl_options": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"certificate": {
+							Type:     schema.TypeString,
+							Optional: false,
+						},
+						"private_key": {
+							Type:     schema.TypeString,
+							Optional: false,
+						},
+						"bundle_method": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+			"hosts": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"issuer": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"signature": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"bundle_method": {
+				Type:     schema.TypeString,
+				Computed: true,
```

Comment:
As the API has restrictions on what values can be used here, it's helpful to provide validation before the HTTP interactions are made.

```suggestion
				Computed: true,
				ValidateFunc: validation.StringInSlice([]string{"ubiquitous", "optimal", "force"}, false),
```

Reply:
We have `bundle_method` here as well as nested under `custom_ssl_options`. Is there a difference for these two?

Reply:
The one under `custom_ssl_options` is passed in during creation. The other is returned by the `/zones/:zone_id/custom_certificates/:cert_id` which the Read method calls. Added it to top level as a convenience since on Read, the `custom_ssl_options` map is not set.

Reply:
🤔 Generally, we've followed what the public API has returned to ensure we had something consistent to map it off. This feels like it will cause some confusion if the value is present in two spots but not matching.

Reply:
I've tried to follow the public API however, the return body for create in the api differs from the request body passed into POST. I've tried to account for both by mapping the POST request body field `custom_ssl_options` as a nested map and then mapping the return body separately where `bundle_method` is non-nested (see below return body `POST zones/:zone_identifier/custom_certificates` from public api).  
```{
  "success": true,
  "errors": [],
  "messages": [],
  "result": {
    "id": "2458ce5a-0c35-4c7f-82c7-8e9487d3ff60",
    "hosts": [
      "example.com"
    ],
    "issuer": "GlobalSign",
    "signature": "SHA256WithRSA",
    "status": "active",
    "bundle_method": "ubiquitous",
    "geo_restrictions": {
      "label": "us"
    },
    "zone_id": "023e105f4ecef8ad9ca31a8372d0c353",
    "uploaded_on": "2014-01-01T05:20:00Z",
    "modified_on": "2014-01-01T05:20:00Z",
    "expires_on": "2016-01-01T05:20:00Z",
    "priority": 1
  }
}```

I think I can set the `bundle_method` on READ within the nested map, `custom_ssl_options` and then remove `bundle_method` from top level in the schema so we don't have it specified in two places. But then the return body won't match 1-1 with our public api. I don't have a strong preference here and want to go with our convention so let me if you are okay with a missing `bundle_method` field in top level of return body.

Reply:
Thanks for the explanation. Would it simply things if we separated this into two structs? One for creating the certificate and one for the API responses?

Reply:
That may work. Are you thinking something like the below ?
```	
return &schema.Resource{
		Create: resourceCloudflareCustomSslCreate,
		Read:   resourceCloudflareCustomSslRead,
		Update: resourceCloudflareCustomSslUpdate,
		Delete: resourceCloudflareCustomSslDelete,
		Importer: &schema.ResourceImporter{
			State: resourceCloudflareCustomSslImport,
		},

		Schema: map[string]*schema.Schema{
			"request": {
				Type: schema.TypeMap,
				........
			},
			"response": {
				Type: schema.TypeMap,
				........
			}
		}
```

Reply:
@jacobbednarz ^

Reply:
Cheers for the poke @jterzis - I've been caught up with other things and hadn't got back to this.

I wouldn't want to change the Terraform schema because the operator doesn't need to know what the underlying struct actually is (for example, why there is a nested section if the data at the top level already matches it). They are just interested in passing in the required data as a structure that makes sense and not so much what the internal library is doing. I was more thinking along the lines of if you needed two different payloads for the POST request/response that there would be a separation of these within `cloudflare-go` instead of lumping them both into one and then having the schema account for both as it is needed. Does that make sense?

Reply:
@jacobbednarz Thanks for adding some clarity to the request. `cloudflare-go` separates the req/resp payloads into 2 different structs for an underlying POST and returns the same response payload on GETS as it does on the POST (https://github.com/cloudflare/cloudflare-go/blob/master/ssl.go#L65-L92) so Im not sure there is much more to do there.

What I am doing in the schema amounts to merging the 2 structs from cloudflare-go (`ZoneCustomSSLOptions`, `ZoneCustomSSL`). In fact, there shouldn't be much confusion in the schema as to which `bundle_method` is needed for CREATING the resource (at time `apply` or `plan`) versus which is READ in since the latter is a `Computed` attribute in the schema.  Maybe it would help if you describe the specific case you are trying to mitigate ?

Reply:
Now that I think about it more, I could probably remove the computed `bundle_method` from the schema since it's not required to be set in READ only on CREATE.

Reply:
> Maybe it would help if you describe the specific case you are trying to mitigate ?

Good idea, let's start with the issue first and work backwards from there 😄 My main concern is that we're duplicating schema fields for the purposes of making the internals happy. That tells me that _something_ isn't encapsulated or structured correctly.

Reply:
Let's abstract away from internals and present the struct that makes sense. And please with no Request/Response fields 🙏

Reply:
Sounds good. I've normalized the schema struct and removed `bundle_method` from the top level struct which was duplicative as @jacobbednarz indicated.